### PR TITLE
Expand SelfHealing.

### PR DESCRIPTION
**Pull request by GraionDilach**
Adds a percentage healing option and damagetypes.
Depends on #11513 (also can be used as a justification of that PR).

Includes testcase.

### CODE REVIEW

**Comment by GraionDilach** [comment]
Rebased, no longer needs dependency.

**Comment by GraionDilach** [comment]
Would it make sense if I'd replace the Dune2000 elite selfhealing with a 4% percentage heal? Considering that even the Rifle Infantry has 600 HP there, the current +2 HP per 125 frames doesn't really make a difference.

File: OpenRA.Mods.Common/Traits/SelfHealing.cs [hunk 1]
```diff
@@ -9,19 +9,30 @@
...
 
 namespace OpenRA.Mods.Common.Traits
 {
 	[Desc("Attach this to actors which should be able to regenerate their health points.")]
 	class SelfHealingInfo : UpgradableTraitInfo, Requires<HealthInfo>
 	{
+		[Desc("Absolute amount of health points added in each step.")]
 		public readonly int Step = 5;
+
+		[Desc("Relative percentages of health added in each step.")]
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Traits/SelfHealing.cs]: In the description, can you mention the fact that the sum of `Step` and `PercentageStep` gets applied when both are defined, please?

**Comment by obrakmann** [comment]
:+1: 

Yeah, go for it. With the current self-healing, some infantry units need almost 20 minutes to heal back to full health from near death (if I didn't screw up my maths). With the 4% option, it would be about two minutes for all units.

File: OpenRA.Mods.Common/Traits/SelfHealing.cs (hunk 1 above)
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/SelfHealing.cs]: Added.

**Comment by GraionDilach** [comment]
Replaced the earlier testcommit with the D2k selfhealing change.

**Comment by pchote** [comment]
Looks fine to me :+1:

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/d1b3edd232e89143477bf2710f43c6f6e46cf924

**Pull request merged**